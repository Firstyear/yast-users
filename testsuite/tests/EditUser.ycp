/**
 * File:
 *  AddGroup.ycp
 *
 * Module:
 *  Users configurator
 *
 * Summary:
 *  Saving group tests.
 *
 * Authors:
 *  Jiri Suchomel <jsuchome@suse.cz>
 *
 */

{
    // testedfiles: Users.pm UsersLDAP.pm UsersCache.pm

    include "testsuite.ycp";
    import "Users";
//    import "UsersCache";
//    import "UsersLDAP";
    import "Mode";
//    import "Progress";
//    import "Report";

    map READ = $[
	"passwd": $[
	    "local": $[
		"users": $[
		    500: $[
			"uid"		: "hh",
			"uidnumber"	: 500,
			"type"		: "local",
			"userpassword"	: "x",
			"sn"		: "HH",
			"grouplist"	: $[
			    "users"	: 1
			],
			"homedirectory"	: "/home/hh",
		    ],
		    "by_name": $[
			"hh":	500,
		    ],
		    "last_uid": 500,
		    "homes":	$[
			"/home/hh":	1,
		    ],
		    "usernames": $[
			"hh":		1,
		    ],
		    "uids": $[
			500:		1,
		    ]
		],
		"shadow": $[
		],
		"groups": $[
		    100: $[
			"cn"	:	"users",
			"gidnumber":	100,
			"type"	:	"local",
			"userlist"	: $[
			    "hh":	1
			],
		    ],
		    "by_name": $[
			"users":	100,
		    ],
		    "gids": $[
			100:		1,
		    ],
		    "groupnames": $[
			"users":	1,
		    ]
		],
	    ],
	    "system": $[
		"users": $[
		    0:	$[
			"uid"		: "root",
			"uidnumber"	: 0
		    ],
		    "by_name": $[
			"root"		: 0
		    ],
		    "last_uid": 0,
		    "homes":	$[
		    ],
		    "usernames": $[
			"root"		: 1,
		    ],
		    "uids": $[
			0		: 1
		    ],
		],
		"shadow": $[
		],
		"groups": $[
		    0: $[
			"cn"		: "root",
			"uidnumber"	: 0,
			"gidnumber"	: 0,
			"type"		: "system"
		    ],
		    "by_name": $[
			"root"	: 0
		    ],
		    "gids": $[
			0	: 1
		    ],
		    "groupnames": $[
			"root"	: 1
		    ]
		],
	    ],
	],
	"etc" : $[
	    "fstab": [],
	    "cryptotab": [],
	    "default": $[
		"useradd": $[
		    "home":	"/home",
		    "groups":	"audio,video",
		    "group":	100
		]
	    ],
	],
	"target": $[
	    "stat"	: $[
		"isdir"	: true
	    ],
	    "size"	: -1,
	    "tmpdir"	: "/tmp/YaST"
	],
	"product": $[
		"features": $[
		    "USE_DESKTOP_SCHEDULER"	: "no",
		    "IO_SCHEDULER"		: "",
		    "ENABLE_AUTOLOGIN"		: "false",
		    "UI_MODE"			: "simple",
		    "EVMS_CONFIG"		: "no"
		]
	],
    ];
    map WRITE = $[];
    map EXEC = $[
	"passwd" : $[
	    "init"	: true
	],
	"target" : $[
	    "bash"		: 0,
	    "bash_output"	: $[],
	],
    ];

    DUMP ("==========================================================");

    Mode::test = true;

    TEST (``(Users::Read ()), [READ, WRITE, EXEC], 0);

    Users::SelectUserByName ("hh");

    DUMP (sformat ("---- user 'hh':\n %1", Users::GetCurrentUser ()));

    map<string,any> changes = $[
	"uidnumber"	: 501,
    ];

    TEST (``(Users::EditUser (changes)), [READ, WRITE, EXEC], 0);

    DUMP (sformat ("---- user 'hh':\n %1", Users::GetCurrentUser ()));

    EXEC ["target", "bash_output", "stdout"] = "hh";

    string error = (string) TEST(``(Users::CheckUser ($[])), [READ,WRITE,EXEC], 0);
    DUMP (sformat ("---- check user after edit:\n %1", error));

    DUMP ("---- commit user:");
    TEST (``(Users::CommitUser ()), [READ, WRITE, EXEC], 0);

    // home changed its owner because of uid change
    TEST (``(Users::Write ()), [READ, WRITE, EXEC], 0);

    Users::SelectUser (500);
    DUMP (sformat ("---- user 500:\n %1", Users::GetCurrentUser ()));

    Users::SelectUser (501);
    DUMP (sformat ("---- user 501:\n %1", Users::GetCurrentUser ()));

    changes = $[
	"homedirectory"	: "/new/home/hh",
	"create_home"	: true,
    ];

    TEST (``(Users::EditUser (changes)), [READ, WRITE, EXEC], 0);

    DUMP (sformat ("---- user 'hh':\n %1", Users::GetCurrentUser ()));

    error = (string) TEST(``(Users::CheckUser ($[])), [READ,WRITE,EXEC], 0);
    DUMP (sformat ("---- check user after edit:\n %1", error));

    DUMP ("---- commit user:");
    TEST (``(Users::CommitUser ()), [READ, WRITE, EXEC], 0);

    // home directory was changed -> move it
    READ ["target", "stat"] = $[];

    TEST (``(Users::Write ()), [READ, WRITE, EXEC], 0);

    DUMP ("==========================================================");
}
