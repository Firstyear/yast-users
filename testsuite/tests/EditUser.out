Dump	==========================================================
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Return	
Return	nil
Dump	---- user 'hh':
Dump	 $["addit_data":"", "cn":"HaHa", "gidNumber":"100", "grouplist":$[], "groupname":"users", "homeDirectory":"/home/hh", "loginShell":"/bin/bash", "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"13727", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "type":"local", "uid":"hh", "uidNumber":"500", "userPassword":"heslo"]
Read	.anyxml "/etc/security/pam_mount.conf.xml" nil
Return	
Dump	---- user 'hh':
Dump	 $["addit_data":"", "chown_home":true, "cn":"HaHa", "create_home":true, "crypted_home_size":"0", "encrypted":true, "gidNumber":"100", "grouplist":$[], "groupname":"users", "homeDirectory":"/home/hh", "loginShell":"/bin/bash", "org_uidNumber":"500", "org_user":$["addit_data":"", "chown_home":true, "cn":"HaHa", "create_home":true, "crypted_home_size":"0", "encrypted":true, "gidNumber":"100", "grouplist":$[], "groupname":"users", "homeDirectory":"/home/hh", "loginShell":"/bin/bash", "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"13727", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "type":"local", "uid":"hh", "uidNumber":"500", "userPassword":"heslo"], "plugins":[], "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"13727", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "type":"local", "uid":"hh", "uidNumber":"501", "userPassword":"heslo", "what":"edit_user"]
Execute	.target.bash_output "echo 'hh' | grep '^[[:alpha:]_][[:alnum:]_.-]*[[:alnum:]_.$-]\\?$'" $["LANG":"C"] $["stdout":"hh"]
Return	
Dump	---- check user after edit:
Dump	 
Dump	---- commit user:
Return	true
Return	nil
Return	nil
Execute	.target.bash "/bin/cp /etc/passwd /etc/passwd.YaST2save" 0
Write	.target.string "/etc/passwd" "at:x:25:25:Batch jobs daemon:/var/spool/atjobs:/bin/bash\nbin:x:1:1:bin:/bin:/bin/bash\ndaemon:x:2:2:Daemon:/sbin:/bin/bash\nmail:x:8:12:Mailer daemon:/var/spool/clientmqueue:/bin/false\nnobody:x:65534:65533:nobody:/var/lib/nobody:/bin/bash\nroot:x:0:0:root:/root:/bin/bash\nuucp:x:10:14:Unix-to-Unix CoPy system:/etc/uucp:/bin/bash\nhh:x:501:100:HaHa:/home/hh:/bin/bash\nii:x:555:100:Test user:/home/ii:/bin/bash\n+::::::\n" true
Execute	.target.bash_output "diff -U 1 /etc/passwd.YaST2save /etc/passwd" $["stdout":"hh"]
Read	.target.stat "/home" $["isdir":true]
Read	.target.stat "/home/hh" $["isdir":true]
Read	.target.stat "/home/hh" $["isdir":true]
Execute	.target.bash_output "/bin/chown -R 501:100 /home/hh" $["stdout":"hh"]
Execute	.target.bash "/bin/cp /etc/shadow /etc/shadow.YaST2save" 0
Write	.target.string "/etc/shadow" "at:!:13636:0:99999:7:::\nbin:*:13636::::::\ndaemon:*:13636::::::\nmail:*:13636::::::\nnobody:*:13636::::::\nroot:password:13636::::::\nuucp:*:13636::::::\nhh:heslo:13727:0:99999:7:::\n+::0:0:0::::\n" true
Execute	.target.bash "/usr/sbin/nscd -i passwd" 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $["custom_groups":["local"], "custom_users":["local"], "dont_warn_when_nisserver_notdes":false, "dont_warn_when_uppercase":false] true
Return	
Return	nil
Dump	---- user 500:
Dump	 $[]
Return	nil
Dump	---- user 501:
Dump	 $["addit_data":"", "chown_home":true, "cn":"HaHa", "create_home":true, "crypted_home_size":"0", "encrypted":true, "gidNumber":"100", "grouplist":$[], "groupname":"users", "homeDirectory":"/home/hh", "loginShell":"/bin/bash", "org_homeDirectory":"/home/hh", "org_uid":"hh", "org_uidNumber":"501", "plugins":[], "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"13727", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "type":"local", "uid":"hh", "uidNumber":"501", "userPassword":"heslo", "what":"edit_user"]
Read	.anyxml "/etc/security/pam_mount.conf.xml" nil
Return	
Dump	---- user 'hh':
Dump	 $["addit_data":"", "chown_home":true, "cn":"HaHa", "create_home":true, "crypted_home_size":"0", "encrypted":true, "gidNumber":"100", "grouplist":$[], "groupname":"users", "homeDirectory":"/new/home/hh", "loginShell":"/bin/bash", "org_homeDirectory":"/home/hh", "org_uid":"hh", "org_uidNumber":"501", "org_user":$["addit_data":"", "chown_home":true, "cn":"HaHa", "create_home":true, "crypted_home_size":"0", "encrypted":true, "gidNumber":"100", "grouplist":$[], "groupname":"users", "homeDirectory":"/home/hh", "loginShell":"/bin/bash", "org_homeDirectory":"/home/hh", "org_uid":"hh", "org_uidNumber":"501", "plugins":[], "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"13727", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "type":"local", "uid":"hh", "uidNumber":"501", "userPassword":"heslo", "what":"edit_user"], "plugins":[], "shadowExpire":"", "shadowFlag":"", "shadowInactive":"", "shadowLastChange":"13727", "shadowMax":"99999", "shadowMin":"0", "shadowWarning":"7", "type":"local", "uid":"hh", "uidNumber":"501", "userPassword":"heslo", "what":"edit_user"]
Execute	.target.bash_output "echo 'hh' | grep '^[[:alpha:]_][[:alnum:]_.-]*[[:alnum:]_.$-]\\?$'" $["LANG":"C"] $["stdout":"hh"]
Return	
Dump	---- check user after edit:
Dump	 
Dump	---- commit user:
Return	true
Execute	.target.bash "/bin/cp /etc/passwd /etc/passwd.YaST2save" 0
Write	.target.string "/etc/passwd" "at:x:25:25:Batch jobs daemon:/var/spool/atjobs:/bin/bash\nbin:x:1:1:bin:/bin:/bin/bash\ndaemon:x:2:2:Daemon:/sbin:/bin/bash\nmail:x:8:12:Mailer daemon:/var/spool/clientmqueue:/bin/false\nnobody:x:65534:65533:nobody:/var/lib/nobody:/bin/bash\nroot:x:0:0:root:/root:/bin/bash\nuucp:x:10:14:Unix-to-Unix CoPy system:/etc/uucp:/bin/bash\nhh:x:501:100:HaHa:/new/home/hh:/bin/bash\nii:x:555:100:Test user:/home/ii:/bin/bash\n+::::::\n" true
Execute	.target.bash_output "diff -U 1 /etc/passwd.YaST2save /etc/passwd" $["stdout":"hh"]
Read	.target.stat "/new/home" $[]
Execute	.target.mkdir "/new/home" 0
Read	.target.stat "/new/home/hh" $[]
Read	.target.stat "/home/hh" $[]
Read	.target.stat "/new/home/hh" $[]
Execute	.target.bash "/bin/cp /etc/shadow /etc/shadow.YaST2save" 0
Write	.target.string "/etc/shadow" "at:!:13636:0:99999:7:::\nbin:*:13636::::::\ndaemon:*:13636::::::\nmail:*:13636::::::\nnobody:*:13636::::::\nroot:password:13636::::::\nuucp:*:13636::::::\nhh:heslo:13727:0:99999:7:::\n+::0:0:0::::\n" true
Execute	.target.bash "/usr/sbin/nscd -i passwd" 0
Return	
Dump	==========================================================
